### PR TITLE
Add operations for scrolling by half a page

### PR DESCRIPTION
**Pull request by dennisschagt**
Resolves https://github.com/newsboat/newsboat/issues/36

Tested using config:
```
bind-key ^U halfpageup
bind-key ^D halfpagedown
```

### CODE REVIEW

File: doc/keycmds.dsv [hunk 1]
```diff
@@ -65,6 +65,8 @@ up||kbd:[UP]||Go up one item in the list.
 down||kbd:[DOWN]||Go down one item in the list.
 pageup||kbd:[PPAGE]||Go up one page in the list.
 pagedown||kbd:[NPAGE]||Go down one page in the list.
+halfpageup||n/a||Go up half a page
+halfpagedown||n/a||Go down half a page
```
Comment (Minoru) [inline doc/keycmds.dsv:69]:
```suggestion
halfpageup||n/a||Go up half a page.
halfpagedown||n/a||Go down half a page.
```

File: test/listmovementcontrol.cpp [hunk 2]
```diff
@@ -393,3 +393,102 @@ TEST_CASE("move_page_down() moves down by `list height` lines, and respects `wra
...
+				list_movement.scroll_halfpage_up();
+
+				REQUIRE(list.position == 0);
+				REQUIRE(list.scroll_offset == 0);
+			}
+		}
+	}
+}
+
+TEST_CASE("scroll_halfpage_down() moves down by `list height` lines, and respects `wrap_scroll`",
```
Comment (Minoru) [inline test/listmovementcontrol.cpp:449]: `wrap_scroll` is not checked in this test, and it's actually what I wondered while reading the implementation: shouldn't these new operations respect `wrap-scroll`? Do you have a rationale for leaving it out?

**Review by Minoru — COMMENTED** [review]
Thank you!

Since I was so slow to review this (for which I'm sorry), there's no time to act about the wrap-scroll question. I hope you won't take a grudge that this won't make it into the upcoming release.

Other than the wrap-scroll support this looks perfect!

File: src/textviewwidget.cpp [hunk 3]
```diff
@@ -103,6 +103,35 @@ void TextviewWidget::scroll_page_down()
...
+		set_scroll_offset(offset - scroll_amount);
+	} else {
+		set_scroll_offset(0);
+	}
+}
+
+void TextviewWidget::scroll_halfpage_down()
+{
+	if (num_lines == 0) {
+		// Ignore if list is empty
```
[the remark this answers is not in the document]
Comment (der-lyse) [reply src/textviewwidget.cpp:124]: I'm wondering: what's the reason that these empty list guards are only used when scrolling down, but left out when scrolling up? I'd expect them to be present there as well, so it's symmetrical.

(I know, you just stuck to the structure that was already there. Addressing this in a separate merge request is absolutely fine for me.)

File: include/listmovementcontrol.h [hunk 4]
```diff
@@ -103,6 +108,50 @@ class ListMovementControl : public Backend {
...
+		} else {
+			current_scroll_offset = max_offset();
+		}
+
+		if (current_position + scroll_amount <= maxpos) {
+			set_position(current_position + scroll_amount);
+		} else {
+			set_position(maxpos);
+		}
+
```
Comment (der-lyse) [inline include/listmovementcontrol.h:152]: I reckon this empty line at the end of the method doesn't add any value regarding readability and hence can be removed. ;-)

File: test/listmovementcontrol.cpp (hunk 2 above)
Comment (dennisschagt) [reply test/listmovementcontrol.cpp:487]: I thought it didn't make sense for this functionality, but don't have a good reason for why I thought that.

File: include/listmovementcontrol.h (hunk 4 above)
Comment (dennisschagt) [reply include/listmovementcontrol.h]: Jup, removed :)

File: doc/keycmds.dsv (hunk 1 above)
Comment (dennisschagt) [reply doc/keycmds.dsv]: Applied + squashed

File: src/textviewwidget.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/textviewwidget.cpp:124]: The reason to check for empty lists when moving down is the calculation of the lowest possible position: `const std::uint32_t maxpos = num_lines - 1;`.
That would wrap around to a very high number if `num_lines` would be 0.

When moving up, we are just interested in not moving futher up than position 0.
That is always possible, even if the list is empty.

**Comment by Minoru** [comment]
Looks great, thank you very much!

File: src/textviewwidget.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (der-lyse) [reply src/textviewwidget.cpp:124]: Oh, didn't know that moving to index 0 is possible for empty lists, thank you for the explanation!

**Pull request merged**